File: packages/trpc/src/schemas.ts
import { z } from 'zod';

export const UserSchema = z.object({
  id: z.string(),
  username: z.string(),
  email: z.string().email(),
  avatarUrl: z.string().nullable(),
  bio: z.string().nullable(),
  location: z.string().nullable(),
  profileUrl: z.string().nullable(),
  createdAt: z.union([z.string(), z.date()]),
  updatedAt: z.union([z.string(), z.date()]),
});

export const MessageSchema = z.object({
  id: z.string(),
  role: z.enum(['user', 'assistant']),
  content: z.string(),
  conversationId: z.string(),
  createdAt: z.union([z.string(), z.date()]),
});

export const ConversationSchema = z.object({
  id: z.string(),
  title: z.string().nullable(),
  userId: z.string(),
  createdAt: z.union([z.string(), z.date()]),
  updatedAt: z.union([z.string(), z.date()]),
});

export const ConversationWithMessagesSchema = ConversationSchema.extend({
  messages: z.array(MessageSchema),
});

export type User = z.infer<typeof UserSchema>;
export type Message = z.infer<typeof MessageSchema>;
export type Conversation = z.infer<typeof ConversationSchema>;
export type ConversationWithMessages = z.infer<typeof ConversationWithMessagesSchema>;

File: apps/api/app/modules/conversations/schemas.py
from datetime import datetime
from enum import Enum

from pydantic import BaseModel, ConfigDict


class MessageRole(str, Enum):
    USER = "user"
    ASSISTANT = "assistant"


class MessageRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    role: MessageRole
    content: str
    parts: list | None = None
    conversation_id: str
    created_at: datetime


class ConversationListRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    title: str | None
    user_id: str
    created_at: datetime
    updated_at: datetime


class ConversationRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    title: str | None
    user_id: str
    created_at: datetime
    updated_at: datetime
    messages: list[MessageRead] = []


class ConversationCreate(BaseModel):
    title: str | None = None


class MessageCreate(BaseModel):
    role: MessageRole
    content: str
    parts: list | None = None


class StreamRequest(BaseModel):
    content: str
